src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -211,14 +215,34 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                     // So, we start with a larger capacity to avoid extra array copies.
                     var results = new List<TagHelperDescriptor>(capacity: 128);
 
+                    var oldCachedDescriptors = _descriptorCache;
+                    var newCachedDescriptors = new Dictionary<MetadataReference, ImmutableArray<TagHelperDescriptor>>(oldCachedDescriptors.Count);
+                    var descriptorsBuilder = new List<TagHelperDescriptor>();
+
                     foreach (var reference in compilation.References)
                     {
-                        if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
+                        if (!oldCachedDescriptors.TryGetValue(reference, out var cachedDescriptors))
                         {
-                            tagHelperFeature.CollectDescriptors(assembly, results);
+                            if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
+                            {
+                                tagHelperFeature.CollectDescriptors(assembly, descriptorsBuilder);
```

Comment:
Note: This calls into the `ITagHelperDescriptorProvider` implementations, which already have caches of `TagHelperProvider[]` per `IAssemblySymbol`.